services/usb-livesync-service-base.ts

```diff
@@ -108,24 +111,40 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
 				gaze("**/*", { cwd: data.watchGlob }, function(err: any, watcher: any) {
 					this.on('all', (event: string, filePath: string) => {
 						if (event === "added" || event === "changed") {
-							if (!that.isFileExcluded(filePath, data.excludedProjectDirsAndFiles, data.projectFilesPath)) {
-								let canExecuteFastLiveSync = data.canExecuteFastLiveSync && data.canExecuteFastLiveSync(filePath);
-
-								if (synciOSSimulator && !canExecuteFastLiveSync) {
-									that.batchSimulatorLiveSync(data, filePath);
+							that.$dispatcher.dispatch(() => (() => {
+								let fileHash = that.$fs.getFileShasum(filePath).wait();
+								if (fileHash === that.fileHashes[filePath]) {
+									that.$logger.trace(`Skipping livesync for ${filePath} file with ${fileHash} hash.`);
+									return;
+								} else {
+									that.$logger.trace(`Adding ${filePath} file with ${fileHash} hash.`);
+									that.fileHashes[filePath] = fileHash;
 								}
 
-								if ((!that.$options.emulator || data.platform.toLowerCase() === "android") && !canExecuteFastLiveSync) {
-									that.batchLiveSync(data, filePath);
-								}
+								if (!that.isFileExcluded(filePath, data.excludedProjectDirsAndFiles, data.projectFilesPath)) {
+									let canExecuteFastLiveSync = data.canExecuteFastLiveSync && data.canExecuteFastLiveSync(filePath);
+
+									if (synciOSSimulator && !canExecuteFastLiveSync) {
+										that.batchSimulatorLiveSync(data, filePath);
+									}
+
+									if ((!that.$options.emulator || data.platform.toLowerCase() === "android") && !canExecuteFastLiveSync) {
+										that.batchLiveSync(data, filePath);
+									}
 
-								if (canExecuteFastLiveSync) {
-									data.fastLiveSync(filePath);
+									if (canExecuteFastLiveSync) {
+										data.fastLiveSync(filePath);
+									}
 								}
-							}
+
+							}).future<void>()());
 						}
 
 						if (event === "deleted") {
+							if (that.fileHashes[filePath]) {
+								delete that.fileHashes[filePath];
```

Comment:
`delete` prevents V8 from optimizing this function, making it 10-20 times slower. Can we avoid it? It is an optional improvement, and I do not insist on it, but if we can do it without making the code too ugly, it is good idea to do it.